### ISSUE TITLE
imfile: multiline log messages can be buffered for a long time

### ISSUE THREAD

**Issue by mhyllander** [rsyslog/rsyslog#1133]
The current behavior of imfile when using the multiline feature is to keep the latest log message in a buffer and appending new lines from the file until the start of a new log message is seen. Only then is the buffered message queued for processing.

While this works OK for log files with a steady flow of log messages, it's a problem when you have log files that are rarely written to. E.g. we have recently turned off debug logging on one of our services, so that basically only warning and error log messages remained. This had the effect that an error log message from one day was not processed until the next day when the next log message was written to the file. Which is not so good.

When I wrote a log collector a couple of years ago I solved this problem by flushing the buffered log message after some time had passed since the last line was read from the file. Basically (re-)setting an alarm after each line is read. If no new lines appear, the alarm will go off, and the buffered log message will be queued for processing.

The amount of time that must pass can of course be configurable, but I think it is rare for multiline log messages to be written one line at a time with long delays in between. In practice all the lines of a multiline log message are written at the same time (e.g. an error message with a stack trace), so the alarm time should be pretty short.

**Comment by mhyllander** [comment]
In addition, I wonder what happens with the buffered log message if rsyslog is restarted? Will it simply be lost, or will it be queue for processing (and possibly saved to a queue file) before rsyslog exist?

**Comment by pocams** [comment]
I'm having the same problem with a new rsyslog deployment.  Files like error logs are where multiline (or readmode 2) is more important, and that's where a new entry needs to be reported quickly.  A timeout of 5 to 10 seconds before a new message is flushed would fix the problem for me.

**Comment by rgerhards** [comment]
Partial messages are stored to the state file, so they are not lost. This requires changes to the way inotify events are handled.

**Comment by davidelang** [comment]
how much overhead is there in setting the alarm each time if you have lots of 
messages arriving? Does this scale if you are monitoring a bunch of different 
files at once.

When I modified imfile to introduce the multiline capibility I was trying to 
keep it simple and so I accepted the delay in processing a message if there were 
no new messages for a while as a way to keep the code simple.

I'm not opposing such a change, just explaining why it is the way it is and 
making sure we don't add unexpected issues while fixing this hole :-)

David Lang
